### ISSUE TITLE
C/C++ Debugger in Eclipse for macOS: (lldb || gdb)

### ISSUE THREAD

**Issue by carrera1** [jhu-en-525-615-fa21/lab1-carrera1#6]
# Problem

Went to Run -> Debug Configurations... -> Debugger Tab

Using following options:

<img width="935" alt="image" src="https://user-images.githubusercontent.com/26331265/131937464-88f4a7c2-a963-48e8-8b0a-5cd579b11c83.png">

Upon pressing "debug" button, receive the following error

<img width="399" alt="image" src="https://user-images.githubusercontent.com/26331265/131937203-1884d528-353e-4c58-9285-2521da61f3df.png">

# Likely Cause

macOS does not come with `gdb` nor `gcc`. In fact, the default `gcc` on macOS is really just `llvm`/`clang` as shown below

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/26331265/131937743-d0eef6cd-5420-4bf7-8f8e-b734bcb06472.png">

# Next Steps

* [ ] Determine if `lldb` can be used with STM32CubeIDE/Eclipse. If so, test out w/ Hello_World_Tutorial and document steps
  * The [CDT User FAQ](https://wiki.eclipse.org/CDT/User/FAQ) seems to have relevant info
  * [8.24 How do I use GDB on recent versions of macOS?](https://wiki.eclipse.org/CDT/User/FAQ#How_do_I_use_GDB_on_recent_versions_of_macOS.3F)
  * [8.25 How do I get the LLDB debugger](https://wiki.eclipse.org/CDT/User/FAQ#How_do_I_get_the_LLDB_debugger.3F)
  * [8.26 How do I install the LLDB debugger integration](https://wiki.eclipse.org/CDT/User/FAQ#How_do_I_install_the_LLDB_debugger_integration.3F)
  * [8.27 How do I debug with LLDB](https://wiki.eclipse.org/CDT/User/FAQ#How_do_I_debug_with_LLDB.3F)
  * [8.28 What are the limitations of using the LLDB debugger in Eclipse](https://wiki.eclipse.org/CDT/User/FAQ#What_are_the_limitations_of_using_the_LLDB_debugger_in_Eclipse.3F)
* [ ] Alternatively... just install `gdb` from source via macports or home-brew, and supply the path to `gdb` in Eclipse
  *  The [CDT User FAQ](https://wiki.eclipse.org/CDT/User/FAQ) seems to imply in previously linked "[8.24 How do I use GDB on recent versions of macOS?](https://wiki.eclipse.org/CDT/User/FAQ#How_do_I_use_GDB_on_recent_versions_of_macOS.3F)", there might be... issues with `gdb` in macOS due to security (code-signing?) reasons

> In recent versions of macOS and Xcode, Apple's GDB is not provided by default. However, it is possible to use the "normal" GDB provided by the Free Software Foundation. For easy installation, you can use package systems such as Brew and MacPorts. Once installed, when you start debugging with GDB you will see an error message such as:
>
> Unable to find Mach task port for process-id 28885: (os/kern) failure (0x5).
 (please check gdb is codesigned - see taskgated(8))
In this case, you need to code sign GDB by following the instructions available at http://sourceware.org/gdb/wiki/BuildingOnDarwin
>
> Note that if GDB is installed with MacPorts, the executable will be named 'ggdb'. You can adjust the preferences in Eclipse to always use this command by going in Preferences > C/C++ > Debug > GDB then change the command to ggdb instead of gdb.
>
> For macOS Sierra (10.12), there seems to be additional problems with GDB that might make it unusable, see the [GDB bug report](https://sourceware.org/bugzilla/show_bug.cgi?id=20266).

NOTE: Appears to be resolved in the reference bug report

_Originally posted by @carrera1 in https://github.com/jhu-en-525-615-fa21/lab1-carrera1/issues/3#issuecomment-912193667_

**Comment by carrera1** [comment]
## What I did
Was able to install LLDB debugger integration by going to Help -> Install New Software

My installed version of Eclipse CDT was 10.2, so I went to the [relevant download page for my CDT version](https://download.eclipse.org/tools/cdt/releases/10.2/) and used its URL to add it as a source

<img width="409" alt="image" src="https://user-images.githubusercontent.com/26331265/131943107-ff615c3b-3a49-4a27-873b-a4a8c87e1a41.png">

And installed the one referenced above in previous comment "8.26 How do I install the LLDB debugger integration" giving a helpful name

## What I'm going to do
Next step is to build `lldb-mi` ("machine interface" used by IDEs, and referenced in CDT User FAQ) since it's no longer bundled with `lldb`

### Notes
* [LLDB MI Driver Info](https://www.codeplay.com/portal/blogs/2014/08/08/lldb-mi-driver-part-1-introduction.html)
* [macOS Developer Forum: lldb-mi no longer bundled with lldb in macOS](https://developer.apple.com/forums/thread/125482)
* [lldb-devel mail list: lldb-mi moved to its own GitHub repo](https://lists.llvm.org/pipermail/lldb-dev/2019-August/015357.html)
  * From the listserv:
    > TL;DR: LLDB no longer provides the lldb-mi executable. You now need to get&compile lldb-mi from the GitHub repository above.

**Comment by carrera1** [comment]
## What I did

* Attempted to build `lldb-mi` to use with Eclipse
  * Numerous build issues related to CMake headaches and ultimately required boiling the LLVM project from scratch just to build `lldb-mi`... hardly worth the time and effort. In other words, I gave up.
* Attempted to build a copy of `gdb` via `macports`
  * Went through the steps to "code sign" the binary so that the Darwin Kernel wouldn't prevent it from hijacking other processes or something to that effect ("System Integrity Protection" on macOS)
  * Testing my own Hello, World program with gdb outside of Eclipse shows... that it's broken. Apparently a common issue.

## What I'm going to do

* Spin up a VM to perform this lab (and possibly the course)
